src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/SafeEvpPKeyHandle.OpenSsl.cs

```diff
@@ -31,13 +38,53 @@ public SafeEvpPKeyHandle(IntPtr handle, bool ownsHandle)
         {
         }
 
+        internal SafeEvpPKeyHandle(IntPtr handle, IntPtr extraHandle)
+            : base(handle, ownsHandle: true)
+        {
+            ExtraHandle = extraHandle;
+        }
+
         protected override bool ReleaseHandle()
         {
-            Interop.Crypto.EvpPkeyDestroy(handle);
+            Interop.Crypto.EvpPkeyDestroy(handle, ExtraHandle);
+            ExtraHandle = IntPtr.Zero;
+
             SetHandle(IntPtr.Zero);
             return true;
         }
 
+        internal static SafeEvpPKeyHandle GenerateRSAKey(int keySize)
+        {
+            return Interop.Crypto.RsaGenerateKey(keySize);
+        }
+
+        internal static SafeEvpPKeyHandle GenerateECKey(int keySize)
+        {
+            SafeEvpPKeyHandle ret = GenerateECKeyCore(new ECOpenSsl(keySize), out int createdKeySize);
+            Debug.Assert(keySize == createdKeySize);
+            return ret;
+        }
+
+        internal static SafeEvpPKeyHandle GenerateECKey(ECCurve curve, out int keySize)
+        {
+            return  GenerateECKeyCore(new ECOpenSsl(curve), out keySize);
+        }
+
+        internal static SafeEvpPKeyHandle GenerateECKey(ECParameters parameters, out int keySize)
+        {
+            return GenerateECKeyCore(new ECOpenSsl(parameters), out keySize);
+        }
+
+        private static SafeEvpPKeyHandle GenerateECKeyCore(ECOpenSsl ecOpenSsl, out int keySize)
+        {
+            using (ECOpenSsl ec = ecOpenSsl)
```

Comment:
It seems really weird to me that this method would dispose an input parameter.  But I guess if it's private maybe it's OK.

Reply:
I added the comment but left mostly as is (except name)